src/shared/utilities/asyncCollection.ts

```diff
@@ -0,0 +1,190 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Coalesce } from './tsUtils'
+
+/**
+ * High-level abstraction over async generator functions of the form `async function*` {@link AsyncGenerator}
+ */
+export interface AsyncCollection<T> extends AsyncIterable<T> {
+    /** Flattens the collection 1-level deep */
+    flatten(): AsyncCollection<SafeUnboxIterable<T>>
+    /** Applies a mapping transform to the output generator */
+    map<U>(fn: (obj: T) => U): AsyncCollection<U>
+    /** Filters out results which will _not_ be passed on to further transformations. */
+    filter<U extends T>(predicate: AsyncPredicate<T, U>): AsyncCollection<U>
+    /** Uses only the first 'count' number of values returned by the generator. */
+    take(count: number): AsyncCollection<T>
+    /** Converts the collection to a Promise, resolving to an array of all values returned by the generator. */
+    promise(): Promise<T[]>
+    /** Converts the collection to a Map, using either a property of the item or a function to select keys. */
+    toMap<K extends StringProperty<T>, U extends string = never>(
+        selector: KeySelector<T, U> | K
+    ): Promise<Map<Coalesce<U, T[K]>, T>>
+    iterator(): AsyncIterator<T, T | void>
+}
+
+/**
+ * Converts an async generator function to an {@link AsyncCollection}
+ *
+ * Uses closures to capture generator functions after each transformation. Generator functions are not called
+ * until a 'final' operation is taken by either:
```

Comment:
- "lazy" would be helpful keyword here. 
- could omit `Uses closures to ...`, that is an implementation detail IIUC. (Also called "thunk"?)

```suggestion
 * Generation is "lazy", that is the generator is not called until needed.
```

The mention of "final operation" might be more confusing than useful--wouldn't it be _more_ surprising if `toCollection` invoked the generator before it's needed?

Reply:
> Wouldn't it be more surprising if toCollection invoked the generator before it's needed?

Yeah it would be more surprising if you understand generators, though I could imagine someone who doesn't understand them being confused why their `map` callback isn't being invoked :)